microsoft-edge/webview2/gettingstarted/wpf.md

```diff
@@ -0,0 +1,291 @@
+{
+    String uri = args.Uri;
+    if (!uri.StartsWith("https://"))
+    {
+        args.Cancel = true;
+    }
+}
+```
+
+In the constructor, we register the EnsureHTTPS EventHandler to the WebView's **NavigationStarting** event. EnsureHTTPS cancels navigation to non-https sites.
+
+Press F5 to build and run the app. Now the app will not navigate to any non-https sites. You can use similar mechanism to accomplish other tasks, such as restricting navigation to within your own domain.
+
+## Step 6 - Scripting
+
+The hosting app can also inject JavaScript into WebView. You can task WebView to execute arbitrary JavaScript or add initialization scripts. Added initialization scripts apply to all future top level document and child frame navigation until removed, and run after the global object has been created and before any other script included by the HTML document is executed.
+
+We can use Scripting to alert the user if they are navigating to a non-https site. Modify the `EnsureHTTPS` function it injects script into the web content using the [ExecuteScriptAsync](../reference/wpf/0-9-494/microsoft-web-webview2-wpf-webview2#executescriptasync) method.
+
+```csharp
+void EnsureHTTPS(object sender, CoreWebView2NavigationStartingEventArgs args)
+{
+    String uri = args.Uri;
+    if (!uri.StartsWith("https://"))
+    {
+        webView.CoreWebView2.ExecuteScriptAsync($"alert('{uri} is not safe, try an https link')");
+        args.Cancel = true;
+    }
+}
+```
+
+Press F5 to build and run the app. Now if you try to navigate to a non-https site, the application alerts you that doing so is not possible.
+
+![https](media/wpf-gettingstarted-https.png)
+
+## Step 7 - Communication between host and web content
+
+The host and the web content can also communicate with each other through `postMessage`. The web content running within a WebView can post to the host through `window.chrome.webview.postMessage`, and the message would be handled by any registered `ICoreWebView2WebMessageReceivedEventHandler` on the host. Likewise, the host can message the web content through `CoreWebView2.PostWebMessageAsString` or `CoreWebView2.PostWebMessageAsJSON`, which would be caught by handlers added from `window.chrome.webview.addEventListener`. The communication mechanism allows the web content to utilize native capabilities by passing messages to ask the host to call native APIs.
+
+As an example to understand the mechanism, when the WebView2 navigates to a URI, lets display that URI in the search bar and alert the user of the URI displayed in the WebView
+
+1. Web Messaging is a **CoreWebView2** functionality. Because the initialization of **CoreWebView2** is asynchronous, create an `InitializeAsync` function that awaits [EnsureCoreWebView2Async](../reference/wpf/0-9-494/microsoft-web-webview2-wpf-webview2#ensurecorewebview2async). In **MainWindow.xaml.cs** update your constructor and define an `InitializeAsync` function:
```

Comment:
```suggestion
1. In **MainWindow.xaml.cs**, update your constructor and create an `InitializeAsync` function as shown in the following code snippet. The `InitializeAsync` function awaits [EnsureCoreWebView2Async](../reference/wpf/0-9-494/microsoft-web-webview2-wpf-webview2#ensurecorewebview2async) because the initialization of `CoreWebView2` is asynchronous.
```